File: examples/server/node/auth/providers/foursquare.js
var express = require('express');
var router = express.Router();
var jwt = require('jwt-simple');
var request = require('request');
var config = require('../../config');
var auth = require('../auth.service');
var User = require('../../api/me/user.model');

/*
 |--------------------------------------------------------------------------
 | Login with Foursquare
 |--------------------------------------------------------------------------
 */
router.post('/auth/foursquare', function(req, res) {
  var accessTokenUrl = 'https://foursquare.com/oauth2/access_token';
  var profileUrl = 'https://api.foursquare.com/v2/users/self';
  var formData = {
    code: req.body.code,
    client_id: req.body.clientId,
    client_secret: config.FOURSQUARE_SECRET,
    redirect_uri: req.body.redirectUri,
    grant_type: 'authorization_code'
  };

  // Step 1. Exchange authorization code for access token.
  request.post({ url: accessTokenUrl, form: formData, json: true }, function(err, response, body) {
    var params = {
      v: '20140806',
      oauth_token: body.access_token
    };

    // Step 2. Retrieve information about the current user.
    request.get({ url: profileUrl, qs: params, json: true }, function(err, response, profile) {
      profile = profile.response.user;

      // Step 3a. Link user accounts.
      if (req.headers.authorization) {
        User.findOne({ foursquare: profile.id }, function(err, existingUser) {
          if (existingUser) {
            return res.status(409).send({ message: 'There is already a Foursquare account that belongs to you' });
          }
          var token = req.headers.authorization.split(' ')[1];
          var payload = jwt.decode(token, config.TOKEN_SECRET);
          User.findById(payload.sub, function(err, user) {
            if (!user) {
              return res.status(400).send({ message: 'User not found' });
            }
            user.foursquare = profile.id;
            user.picture = user.picture || profile.photo.prefix + '300x300' + profile.photo.suffix;
            user.displayName = user.displayName || profile.firstName + ' ' + profile.lastName;
            user.save(function() {
              var token = auth.createJWT(user);
              res.send({ token: token });
            });
          });
        });
      } else {
        // Step 3b. Create a new user account or return an existing one.
        User.findOne({ foursquare: profile.id }, function(err, existingUser) {
          if (existingUser) {
            var token = auth.createJWT(existingUser);
            return res.send({ token: token });
          }
          var user = new User();
          user.foursquare = profile.id;
          user.picture = profile.photo.prefix + '300x300' + profile.photo.suffix;
          user.displayName = profile.firstName + ' ' + profile.lastName;
          user.save(function() {
            var token = auth.createJWT(user);
            res.send({ token: token });
          });
        });
      }
    });
  });
});

module.exports = router;

File: examples/server/ruby/app/models/oauth/foursquare.rb
module Oauth
  class Foursquare < Oauth::Base
    ACCESS_TOKEN_URL = 'https://foursquare.com/oauth2/access_token'
    USER_URL = 'https://api.foursquare.com/v2/users/self'

    def get_data
      response = @client.get(USER_URL, oauth_token: @access_token, v: 20140806)
      @data = JSON.parse(response.body)['response']['user']
      @uid = @data["id"]
      @data
    end

    def formatted_user_data
      {
        provider:       'foursquare',
        token:          @access_token,
        uid:            @data['id'],
        first_name:     @data['firstName'],
        last_name:      @data['lastName'],
        email:          @data['contact']['email'],
        image_url:      "#{@data['photo']['prefix']}256x256#{@data['photo']['suffix']}",
        foursquare_profile: "https://foursquare.com/user/#{@data['id']}"
      }
    end

  end
end


